### ISSUE TITLE
Please disallow migration from Xen x.y to x.y-1 hypervisors unless forced

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1001]
Originally reported of Google Code with ID 945.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

# gnt-cluster --version; gnt-cluster version; hspace --version
gnt-cluster (ganeti v2.10.1) 2.10.1
Software version: 2.10.1
Internode protocol: 2100000
Configuration format: 2100000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version v2.10.1
hspace (ganeti) version v2.10.1
compiled with ghc 7.4
running on linux x86_64

<b>What distribution are you using?</b>

Ubuntu Precise.

<b>What steps will reproduce the problem?</b>
1. Migrate instance between node with Xen 4.x and Xen 4.x-1

<b>What is the expected output? What do you see instead?</b>

Expected output: Job is rejected because the instance will most likely not survive the migration (http://wiki.xen.org/wiki/Xen_Project_Version_Compatibility) and no option to disable that safety-check was set (as it doesn't exist).

Result: The job log contains the following message:
Mon Sep 15 12:42:27 2014 * warning: hypervisor version mismatch between source ([4, x]) and target ([4, x-1]) node

After which the instances crashes on the target node.

<b>Please provide any additional information below.</b>
```
Originally added on 2014-09-15 14:02:36 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Assigned to pudlak to confirm version, this could also be treated as a defect.
```
Originally added on 2014-09-15 14:08:39 +0000 UTC.

Added Labels: Type-Enhancement Priority-High
Changed State: Accepted
Added to Milestone: Release2.12

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-09-25 11:15:05 +0000 UTC.

Added Labels: Priority-Critical Component-htools
Removed Labels: Priority-High
Added to Milestone: Release2.14
Removed from Milestone: Release2.12

**Comment by Ganeti-Issues-Migrator** [comment]
```
Related:

commit 13cb78805d1e3cf15c636c963bda7f89ebe9a655
Author: Klaus Aehlig <aehlig@google.com>
Date:   Thu Sep 25 15:57:48 2014 +0200

    Design: also support location based migration restrictions
    
    Extend the design of location awareness by supporting migration restriction.
    In this way, it will be easier to keep the cluster balanced in a situation
    where not all migrations are possible, e.g., while upgrading a hypervisor.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

```
Originally added on 2014-09-25 15:12:00 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Isse #966 will track the htools related changes
```
Originally added on 2014-10-02 09:40:18 +0000 UTC.

Removed Labels: Component-htools
Added to Milestone: Release2.13
Removed from Milestone: Release2.14

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit a0c3c610991b81a0ffcff4d20776ba08b594905a
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 29 17:17:59 2014 +0200

    Document HV version checks in NEWS
    
    The hypervisor-version check introduced an incompatible change:
    Ganeti now errors out in situations where it used to warn and
    then try and hope for the best. So add a not in the section on
    incompatible changes.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit af18b002f6a0e81a5f03004d124f5b2cec4eb2fd
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 29 17:02:09 2014 +0200

    Add --ignore-hvversions to gnt-instance migrate man page
    
    Describe the new option and the changed default behaviour
    in the man page.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 457c8c8ce1bded9a40591c9ffe6415b04908bc95
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 29 16:19:13 2014 +0200

    Add --ignore-hvversions to gnt-instance migrate
    
    ...which simply passes on the option.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 5a03bfb027b663c42027adc7f5c79d1584bbceb2
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 29 16:11:43 2014 +0200

    Add --ignore-hvversions option
    
    This option will be used to tell Ganeti to not error out if a
    migration between nodes with incompatible hypervisor versions
    is attempted.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit f42fdcd140d9125edbf22609ca94a62502a767f1
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 29 15:13:06 2014 +0200

    Honor ignore_hvversions in LUInstanceMigrate
    
    ...by passing the parameter also to the tasklet.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 270cb9b8061b74f2faf2855fcdb9d384d9ef0e3f
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 29 11:29:05 2014 +0200

    Add pIgnoreHVVersions to OpInstanceMigrate
    
    As a parameter with default value, it will be backwards
    compatible on the wire. Inside our code, however, we have
    to adapt the various calls to the constructor.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 4c943bceec429d7c75e4c810c036639c8e73c304
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 29 11:28:19 2014 +0200

    Add opcode Parameter pIgnoreHVVersion
    
    This parameter will instruct instance migration to proceed in the presence
    of incompatible hypervisor versions.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 3d5e5214abcee2e144599e1fe2e2d7ab652f2808
Author: Klaus Aehlig <aehlig@google.com>
Date:   Fri Sep 26 17:29:33 2014 +0200

    Make hypervisor check fatal
    
    ...unless force is used. This will prevent accidentally
    crashing instances.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit bf488717da2aeb843907fb63f2d1d5e423481a1d
Author: Klaus Aehlig <aehlig@google.com>
Date:   Fri Sep 26 16:44:54 2014 +0200

    Add unit test for HVVersionsLikelySafeForMigration
    
    ...by looking at the results of a few typical examples.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 38049d0149dd0a92720faa7a5c13816c6a121c16
Author: Klaus Aehlig <aehlig@google.com>
Date:   Thu Sep 25 18:27:46 2014 +0200

    Add a function comparing hypervisor versions
    
    The relevance of a warning about migrating between hypervisors
    of different versions depends on whether the migration is likely
    to succeed or not. While it is hard to predict this in general,
    this function follows best practices for Xen's current numbering
    scheme.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

```
Originally added on 2014-10-02 09:41:01 +0000 UTC.

Added Labels: Type-Defect
Removed Labels: Type-Enhancement
Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Reopening as the current solution breaks kvm; see discussion
on the patch series after its push

https://groups.google.com/forum/#!topic/ganeti-devel/0nBVNSzYCa4
```
Originally added on 2014-10-02 11:06:12 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
I'll try to have a look
```
Originally added on 2014-10-06 15:18:49 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-10-07 10:40:28 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 2c5557b1177c1e59d7057911c3e9e7617c7dc06b
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Oct 7 10:48:04 2014 +0200

    Be optimistic for kvm migration
    
    Already in Ganeti 2.8, the machine_version feature was added to
    the kvm hypervisor which allows migration in most cases. In order
    not to block legitimate use cases, restore the old behavior and
    do not check for version on kvm.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit 6a01a1bc189598084ae3dd8f3f324b323ebb2aa2
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Oct 7 10:20:25 2014 +0200

    Move Xen-specific version test hv_xen
    
    The hypervisor-compatibility test we used was specific to
    Xen, so move it to the hv_xen instead of having it in util.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit 00bf6a1dedc72c79292951e23b93d340918a0cc1
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Oct 6 18:23:21 2014 +0200

    Use HV-specific version comparison function
    
    ...to decide if a migration is safe.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit 6a8afdb1cca2187d0e8631fafc597bf8234c3c29
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Oct 6 17:57:18 2014 +0200

    Extend the Hypervisor interface by VersionsSafeForMigration
    
    The question which hypervisor versions are safe to migrate between
    is hypervisor dependent. The default implementation is that it's
    probably fine if the versions are equal and otherwise be safe and
    disallow it.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

```
Originally added on 2014-10-07 10:40:45 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
This restored the old behavior for kvm while
keeping the requested behavior for Xen (check
versions unless override option given). Having
checks for kvm is a different issue; feel free
to open one, if needed.
```
Originally added on 2014-10-07 10:42:19 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:03 +0000 UTC.

Changed State: Released
